worlds/sc2/pool_filter.py

```diff
@@ -88,397 +178,191 @@ def attempt_removal(item: StarcraftItem) -> bool:
-                        if parent_item in inventory:
+            if min_upgrades_per_unit > 0:
+                parent_item = child_to_main_parent.get(item.name)
+                if parent_item is not None:
+                    children = self.item_name_to_child_items.get(parent_item.name, [])
+                    children = [x for x in children if not (ItemFilterFlags.CulledOrBetter & x.filter_flags)]
+                    count = len(children)
+                    if count <= min_upgrades_per_unit:
+                        if parent_item in removable:
                             # Attempt to remove parent instead, if possible
                             item = parent_item
                         else:
                             # Lock remaining upgrades
-                            for item in self.item_children[parent_item]:
-                                if item in inventory:
-                                    inventory.remove(item)
-                                    locked_items.append(copy_item(item))
-                            continue
-
-            # Drop child items when removing a parent
-            if item in parent_items:
-                items_to_remove = [item for item in self.item_children[item] if item in inventory]
-                success = attempt_removal(item)
-                if success:
-                    while len(items_to_remove) > 0:
-                        item_to_remove = items_to_remove.pop()
-                        if item_to_remove not in inventory:
-                            continue
-                        attempt_removal(item_to_remove)
-            else:
-                # Unimportant upgrades may be added again in the second pass
-                if attempt_removal(item):
-                    unused_items.append(item.name)
-
-        pool_items: List[str] = [item.name for item in (inventory + locked_items + self.existing_items + necessary_items)]
-        unused_items = [
-            unused_item for unused_item in unused_items
-            if item_list[unused_item].parent_item is None
-               or item_list[unused_item].parent_item in pool_items
-        ]
+                            for item in children:
+                                item.filter_flags |= ItemFilterFlags.Locked
+                            return False
+            if attempt_removal(item, remove_flag):
+                remove_child_items(item, remove_flag)
+                return True
+            return False
+
+        def item_included(item: StarcraftItem) -> bool:
+            return (
+                ItemFilterFlags.Removed not in item.filter_flags
+                or (ItemFilterFlags.Unexcludable & item.filter_flags)
+                or (not (ItemFilterFlags.Excluded & item.filter_flags)
+                    and (ItemFilterFlags.Requested & item.filter_flags)
+                )
+                or not (ItemFilterFlags.Culled & item.filter_flags)
+            ) != 0
```

Comment:
All I can see there is an expression returning boolean compared to an integer

Reply:
This is mypy typing stuff. Expressions with `not` return a boolean, expressions with `in` or `&` actually return an `ItemFilterFlags`. `X and Y` doesn't return bool, it returns `Type[X] | Type[y]`, so the type of the whole bracketed expression is `bool | ItemFilterFlags`.

Both of those are subclasses of int (bool in Python is a special case of int that is always 0 or 1). So != 0 just casts it to a bool.

I can change it to `bool(expr)` instead; I was getting mixed up with JS for a moment where `Boolean(false)` is truthy because that language is insane.

Reply:
Either way this looks weird, as you use an integer comparison on something that's a result of boolean operators (multiple `or`s connecting `&` and `not in` between flags)

Reply:
Changed to `bool(expr)` instead of `(expr) != 0